docs/calendar-draft.md

```diff
@@ -0,0 +1,315 @@
+
+## Temporal.Date internal slots
+
+Temporal.Date currently has three internal slots: year, month, and day. (An "internal slot" refers to actual data, as opposed to "properties", which could be computed.)  In this proposal, three more internal slots would be added:
+
+1. calendar
+2. isLeapMonth
+3. era
+
+It is believed that a date in any known calendar can be described using the five-tuple (*year*, *month*, *day*, *isLeapMonth*, *era*).  Since all operations involving the slots of Temporal.Date will go through the calendar first, calendars that need more data can assign complex types to these slots.  It is possible that *isLeapMonth* can be merged into *month* by expanding the set of numbers that *month* can represent.
+
+The *calendar* slot contains an object implementing the Temporal.Calendar interface, described below.
+
+## Temporal.Calendar interface
+
+The new Temporal.Calendar interface is a mechanism to allow arbitrary calendar systems to be implemented on top of Temporal.  ***Most users will not encounter the Temporal.Calendar interface directly***, unless they are building or using a non-built-in calendar system.
+
+### Methods on the Temporal.Calendar interface
+
+All of the following methods return new Temporal.Date objects.  All properties of `Temporal.Calendar` are Symbols.
+
+```javascript
+class MyCalendar {
+
+	///////////////////
+	//  To/From ISO  //
+	///////////////////
+
+	/** Returns the projection of self in the ISO calendar */
+	[Temporal.Calendar.toISO](
+		self: Temporal.Date
+	) : Temporal.Date;
+
+	/** Returns the projection of isoDate in the custom calendar */
+	[Temporal.Calendar.fromISO](
+		isoDate: Temporal.Date
+	) : Temporal.Date;
+
+	/** A string identifier for this calendar */
+	[Temporal.Calendar.id] : string;
+
+	//////////////////
+	//  Arithmetic  //
+	//////////////////
+
+	/** Returns self plus duration according to the calendar rules. */
+	[Temporal.Calendar.plus](
+		self: Temporal.Date,
+		duration: Temporal.Duration,
+		options: /* options bag */
+	) : Temporal.Date;
+
+	/** Returns self minus duration according to the calendar rules. */
+	[Temporal.Calendar.minus](
+		self: Temporal.Date,
+		duration: Temporal.Duration,
+		options: /* options bag */
+	) : Temporal.Date;
+
+	/** Returns self minus other, which are dates in the same calendar. */
+	[Temporal.Calendar.difference](
+		self: Temporal.Date,
+		other: Temporal.Date,
+		options: /* options bag */
+	) : Temporal.Duration;
+
+	////////////////////////////////////
+	//  Accessors:                    //
+	//  Semantics defined in date.md  //
+	////////////////////////////////////
+
+	[Temporal.Calendar.dayOfWeek](
+		self: Temporal.Date
+	) : number;
+
+	[Temporal.Calendar.weekOfYear](
+		self: Temporal.Date
+	) : number;
+
+	[Temporal.Calendar.daysInMonth](
+		self: Temporal.Date
+	) : number;
+
+	[Temporal.Calendar.daysInYear](
+		self: Temporal.Date
+	) : number;
+
+	[Temporal.Calendar.leapYear](
+		self: Temporal.Date
+	) : number;
```

Comment:
The return type should be `boolean` instead of `number`.

Reply:
I wonder if we should rename this `yearProperties` and make it return a calendar-specific object.  For example, in iso/gregorian, there are two "types" of years: leap years and non-leap years.  But in Hebrew, there are several more types: not only leap years and non-leap years, but also "kesidran", "chaser", "maleh".  The Hebrew calendar could choose to return a `{ isLeapYear, yearType }` record, while other calendars could return `{ isLeapYear }`.

Reply:
Should we lump all of the computed properties (not only `isLeapYear` but also `weekOfYear`, etc) into a calendar-specific object?  For example (name bikeshedding aside):

```javascript
// Current
d.weekOfYear
d.isLeapYear

// New
d.dateProperties.weekOfYear
d.dateProperties.isLeapYear
```

@pipobscure

Reply:
Splitting this to https://github.com/tc39/proposal-temporal/issues/291 and adding a reference to the explainer.